File: Treap.js
class BSTNode {
  priority;
  value;
  count = 1;
  left = null;
  right = null;
  constructor(value = 0) {
    this.priority = Math.random();
    this.value = value;
  }
}

class Treap {
  size = 0;
  root = null;

  constructor() {}

  split(node, key) {
    if (!node) return [null, null];
    if (node.value < key) {
      const [lesser, greater] = this.split(node.right, key);
      node.right = lesser;
      return [node, greater]
    } else {
      const [lesser, greater] = this.split(node.left, key);
      node.left = greater;
      return [lesser, node];
    }
  }

  merge(node1, node2) {
    if (!node1  || !node2) return node1 || node2;
    if (node1.priority > node2.priority) {
      node1.right = this.merge(node1.right, node2);
      return node1;
    } else {
      node2.left = this.merge(node1, node2.left);
      return node2;
    }
  }

  add(value) {
    let newNode;
    let [lesser, greater] = this.split(this.root, value);
    let equal;
    [equal, greater] = this.split(greater, value + 1);
    if (equal) {
      equal.count++;
      newNode = equal;
    } else {
      newNode = new BSTNode(value);
    }
    lesser = this.merge(lesser, newNode);
    this.root = this.merge(lesser, greater);
    this.size++;
  }

  delete(value) {
    let [lesser, greater] = this.split(this.root, value);
    let equal = null;
    [equal, greater] = this.split(greater, value + 1);
    if (equal) {
      if (equal.count > 1) {
        equal.count--;
        lesser = this.merge(lesser, equal);
      }
      this.size--;
    }
    this.root = this.merge(lesser, greater);
  }

  has(key) {
    let tmp = this.root;
    while (tmp) {
      if (key < tmp.value) {
        tmp = tmp.left;
      } else if (tmp.value === key) {
        return true;
      } else {
        tmp = tmp.right;
      }
    }
    return false;
  }

  prevValue(key) {
    const [lesser, greater] = this.split(this.root, key);
    let tmp = lesser;
    while (tmp && tmp.right) {
      tmp = tmp.right;
    }
    this.root = this.merge(lesser, greater)
    return tmp ? tmp.value : null;
  }

  nextValue(key) {
    const [lesser, greater] = this.split(this.root, key + 1);
    let tmp = greater;
    while (tmp && tmp.left) {
      tmp = tmp.left;
    }
    this.root = this.merge(lesser, greater)
    return tmp ? tmp.value : null;
  }

  min() {
    let tmp = this.root;
    while (tmp && tmp.left) {
      tmp = tmp.left;
    }
    return tmp ? tmp.value : null;
  }

  max() {
    let tmp = this.root;
    while (tmp && tmp.right) {
      tmp = tmp.right;
    }
    return tmp ? tmp.value : null;
  }

  print() {
    let tmp = this.root;
    const stack = [];
    while (stack.length || tmp) {
      while (tmp) {
        stack.push(tmp);
        tmp = tmp.left;
      }
      tmp = stack.pop();
      console.log(tmp.value, tmp.count);
      tmp = tmp.right;
    }
  }

  indexOf(key) {
    let tmp = this.root;
    let index = 0;
    const stack = [];
    while (stack.length || tmp) {
      while (tmp) {
        stack.push(tmp);
        tmp = tmp.left;
      }
      tmp = stack.pop();
      if (tmp.value === key) return index;
      index++;
      tmp = tmp.right;
    }
    return -1;
  }

  getByIndex(i) {
    let tmp = this.root;
    let index = 0;
    const stack = [];
    while (stack.length || tmp) {
      while (tmp) {
        stack.push(tmp);
        tmp = tmp.left;
      }
      tmp = stack.pop();
      if (index === i) return tmp.value;
      index++;
      tmp = tmp.right;
    }
    return null;
  }
}

File: go/treap/treap.go
package treap

type Node struct {
	priority int
	val      int
	count    int
	l        *Node
	r        *Node
}

type Treap struct {
	Root *Node
	Seed int
}

func merge(a, b *Node) *Node {
	if a == nil {
		return b
	}
	if b == nil {
		return a
	}
	if a.priority > b.priority {
		a.r = merge(a.r, b)
		return a
	} else {
		b.l = merge(a, b.l)
		return b
	}
}

func split(key int, node *Node) (l *Node, r *Node) {
	if node == nil {
		return
	}
	if node.val < key {
		l = node
		l.r, r = split(key, node.r)
	} else {
		r = node
		l, r.l = split(key, node.l)
	}
	return
}

func (t *Treap) Add(val int) {
	var l, equal, r *Node
	l, r = split(val, t.Root)
	equal, r = split(val+1, r)
	if equal != nil {
		equal.count++
	} else {
		equal = t.MakeNode(val)
	}
	t.Root = merge(l, merge(equal, r))
}

func (t *Treap) Delete(val int) {
	var l, equal, r *Node
	l, r = split(val, t.Root)
	equal, r = split(val+1, r)
	if equal != nil && equal.count > 1 {
		equal.count--
		t.Root = merge(l, merge(equal, r))
		return
	}
	t.Root = merge(l, r)
}

func (t *Treap) Has(val int) bool {
	node := t.Root
	for node != nil {
		if node.val == val {
			return true
		}
		if val < node.val {
			node = node.l
		} else {
			node = node.r
		}
	}
	return false
}

func (t *Treap) MakeNode(val int) *Node {
	return &Node{
		priority: t.Rand(),
		val:      val,
		count:    1,
	}
}

func (t *Treap) Lower(val int) int {
	l, r := split(val+1, t.Root)
	if l == nil {
		return -1
	}
	node := l
	for node.r != nil {
		node = node.r
	}
	merge(l, r)
	return node.val
}

func (t *Treap) Upper(val int) int {
	l, r := split(val, t.Root)
	if r == nil {
		return -1
	}
	node := r
	for node.l != nil {
		node = node.l
	}
	merge(l, r)
	return node.val
}

func (t *Treap) Inorder(node *Node) {
	if node == nil {
		return
	}
	t.Inorder(node.l)
	t.Inorder(node.r)
}

func (t *Treap) Rand() int {
	t.Seed = (1103515245*t.Seed + 12345) % 2147483648
	return t.Seed
}
